craft_parts/parts.py

```diff
@@ -16,15 +16,137 @@
 
 """Definitions and helpers to handle parts."""
 
-import copy
+from dataclasses import dataclass
 from pathlib import Path
 from typing import Any, Dict, List, Optional, Sequence, Set
 
-from craft_parts import errors
+from craft_parts import errors, unmarshal
 from craft_parts.dirs import ProjectDirs
 from craft_parts.steps import Step
 
 
+@dataclass(frozen=True)
+class PartSpec:
+    """The part specification data."""
+
+    plugin: Optional[str]
+    source: Optional[str]
+    source_checksum: str
+    source_branch: str
+    source_commit: str
+    source_depth: int
+    source_subdir: str
+    source_tag: str
+    source_type: str
+    disable_parallel: bool
+    after: List[str]
+    stage_snaps: List[str]
+    stage_packages: List[str]
+    build_snaps: List[str]
+    build_packages: List[str]
+    build_environment: List[Dict[str, str]]
+    build_attributes: List[str]
+    organize_fileset: Dict[str, str]
+    stage_fileset: List[str]
+    prime_fileset: List[str]
+    override_pull: Optional[str]
+    override_build: Optional[str]
+    override_stage: Optional[str]
+    override_prime: Optional[str]
+
+    @classmethod
+    def unmarshal(cls, data: Dict[str, Any]) -> "PartSpec":
+        """Create and populate a new ``PartSpec`` object from dictionary data.
+
+        The unmarshal method validates and consumes entries in the input
+        dictionary, populating the corresponding fields in the data object.
+
+        :param data: The dictionary data to unmarshal and consume.
+
+        :return: The newly created object.
+
+        :raise ValueError: If data validation fails.
+        """
+        if not isinstance(data, dict):
+            raise ValueError("part data is not a dictionary")
+
+        udata = unmarshal.DataUnmarshaler(data)
+
+        return cls(
+            plugin=udata.pop_optional_string("plugin"),
```

Comment:
I will say that seeing pop in unmarshal is a bit odd.  From a library level, the passed in dict is going to be deepcopy()'d before craft-parts modifies it, right?

Reply:
And the popping is just to ensure there are no extra fields after Parts and Plugin?

If the plugin properties were stripped first (i.e. `plugin: make`  filters out `make-*`, then unmarshal can sanity check by doing its own `data.copy()` and check for leftover fields before returning.

Reply:
Changing to make data consumption optional (by passing `consume=True` in `DataUnmarshaler` constructor) and rename `pop_*` methods to `get_*` to make the utility more generic, implementing the regular non-consuming case as the default. Then use the special consuming case in part specification unmarshal to detect if fields left without changing caller data, raising an error if unexpected fields found. Reorganize tests.